pkg/ble/device/twindata.go

```diff
@@ -0,0 +1,127 @@
+package device
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/currantlabs/ble"
+	"github.com/kubeedge/kubeedge/cloud/pkg/apis/devices/v1alpha2"
+	"github.com/kubeedge/mappers-go/pkg/ble/configmap"
+	"github.com/kubeedge/mappers-go/pkg/ble/driver"
+	"github.com/kubeedge/mappers-go/pkg/ble/globals"
+	"github.com/kubeedge/mappers-go/pkg/common"
+	"k8s.io/klog/v2"
+)
+
+// TwinData is the timer structure for getting twin/data.
+type TwinData struct {
+	BluetoothClient        *driver.BluetoothClient
+	Name                   string
+	Type                   string
+	BluetoothVisitorConfig configmap.BluetoothVisitorConfig
+	Result                 string
+	Topic                  string
+}
+
+// Run timer function.
+func (td *TwinData) Run() {
+	uuid := ble.MustParse(td.BluetoothVisitorConfig.CharacteristicUUID)
+	if p, err := td.BluetoothClient.Client.DiscoverProfile(true); err == nil {
+		if u := p.Find(ble.NewCharacteristic(uuid)); u != nil {
+			c := u.(*ble.Characteristic)
+			// read data actively
+			if (c.Property & ble.CharRead) != 0 {
+				b, err := td.BluetoothClient.Client.ReadCharacteristic(c)
+				if err != nil {
+					klog.Errorf("Failed to read characteristic: %s\n", err)
+					return
+				}
+				td.Result = fmt.Sprintf("%f", td.ConvertReadData(b))
+				if err = td.handlerPublish(); err != nil {
+					klog.Error("publish data to mq failed")
+					return
+				}
+			}
+			// If this Characteristic suports notifications and there's a CCCD
+			// Then subscribe to it
+			if (c.Property&ble.CharNotify) != 0 && c.CCCD != nil {
+				if err := td.BluetoothClient.Client.Subscribe(c, false, td.notificationHandler()); err != nil {
```

Comment:
Not clear why there's a subscription. This function will be called periodically. If the period is small, then there will be lots of subscribing and publishing.

Reply:
well,because bluetooth support two method to fetch data,one is read and another is notify,thers is no clearly defination in crd,so I don't know it's read or notify.